dvc/remote/base.py

```diff
@@ -818,14 +824,17 @@ def _checkout_dir(
         for entry in dir_info:
             relative_path = entry[self.PARAM_RELPATH]
             entry_checksum = entry[self.PARAM_CHECKSUM]
+            entry_cache_info = self.checksum_to_path_info(entry_checksum)
             entry_info = path_info / relative_path
-            self._checkout_file(
-                entry_info,
-                entry_checksum,
-                force,
-                progress_callback,
-                save_link=False,
-            )
+
+            entry_checksum_info = {self.PARAM_CHECKSUM: entry_checksum}
+            if self.changed(entry_info, entry_checksum_info):
+                if self.exists(entry_info):
+                    self.safe_remove(entry_info, force=force)
+                self.link(entry_cache_info, entry_info)
+                self.state.save(entry_info, entry_checksum)
```

Comment:
Do we need to re-introduce separate checkout logic for file and for dir?
I believe we could use `_checkout_file` with no progress callback here. Saving link in case of single file checkout can be handled by some boolean argument, or in `_checkout`. What do you think?

Reply:
We are effectively make both worse) And Boolean flag is not a good sign either.